### PR TITLE
Run main progress bar independent of val progress bar in `TQDMProgressBar`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes some issues with tqdm progress bar. Explained in comments

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @rohitgr7

### CODE REVIEW

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 1]
```diff
@@ -268,8 +268,9 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", *_: Any) -> None:
         self.main_progress_bar.set_description(f"Epoch {trainer.current_epoch}")
 
     def on_train_batch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", *_: Any) -> None:
-        if self._should_update(self.train_batch_idx, self.total_train_batches):
-            _update_n(self.main_progress_bar, self.train_batch_idx + self._val_processed)
+        current = self.train_batch_idx + self._val_processed
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:271]: here we need to track the current index for the main progress bar. With `val_check_interval`, this was incorrect. Added a test that fails on master.

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 2]
```diff
@@ -294,10 +295,12 @@ def on_validation_batch_start(
         self.val_progress_bar.set_description(f"{desc} DataLoader {dataloader_idx}")
 
     def on_validation_batch_end(self, trainer: "pl.Trainer", *_: Any) -> None:
-        if self._should_update(self.val_batch_idx, self.total_val_batches_current_dataloader):
+        if self._should_update(self.val_batch_idx, self.val_progress_bar.total):
             _update_n(self.val_progress_bar, self.val_batch_idx)
-            if trainer.state.fn == "fit":
-                _update_n(self.main_progress_bar, self.train_batch_idx + self._val_processed)
+
+        current = self.train_batch_idx + self._val_processed
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:301]: here, the `main progress bar` now updates independently of `val progress bar`. So in case `val progress bar` doesn't need an update but the `main progress bar` does, it will update.

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 3]
```diff
@@ -338,7 +341,7 @@ def on_predict_batch_start(
         self.predict_progress_bar.set_description(f"{self.predict_description} DataLoader {dataloader_idx}")
 
     def on_predict_batch_end(self, *_: Any) -> None:
-        if self._should_update(self.predict_batch_idx, self.total_predict_batches_current_dataloader):
+        if self._should_update(self.predict_batch_idx, self.predict_progress_bar.total):
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:344]: we assign `total` to the progress bar already, so we can extract it from there itself. Just makes it less error-prone.

File: pytorch_lightning/callbacks/progress/tqdm_progress.py [hunk 4]
```diff
@@ -318,7 +321,7 @@ def on_test_batch_start(
         self.test_progress_bar.set_description(f"{self.test_description} DataLoader {dataloader_idx}")
 
     def on_test_batch_end(self, *_: Any) -> None:
-        if self._should_update(self.test_batch_idx, self.total_test_batches_current_dataloader):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/progress/tqdm_progress.py:321]: All these `self.total_*` properties are still used by Rich, correct?

Does that mean Rich also has this problem?

File: tests/callbacks/test_tqdm_progress_bar.py [hunk 5]
```diff
@@ -347,10 +348,10 @@ def test_tqdm_progress_bar_value_on_colab(tmpdir):
         [2, 3, 1, [1, 2, 3, 4, 5], [1, 2, 3]],
         [0, 0, 3, None, None],
         [1, 0, 3, [1], None],
-        [1, 1, 3, [1, 2], [1]],
```
Comment (carmocca) [inline tests/callbacks/test_tqdm_progress_bar.py:350]: Can you explain why this changed?

File: tests/callbacks/test_tqdm_progress_bar.py [hunk 6]
```diff
@@ -347,10 +348,10 @@ def test_tqdm_progress_bar_value_on_colab(tmpdir):
...
+        [1, 1, 3, [2], [1]],
         [5, 0, 3, [3, 5], None],
-        [5, 2, 3, [3, 5, 7], [2]],
-        [5, 2, 6, [5, 7], [2]],
```
Comment (carmocca) [inline tests/callbacks/test_tqdm_progress_bar.py:353]: Same

File: tests/callbacks/test_tqdm_progress_bar.py [hunk 7]
```diff
@@ -549,16 +550,56 @@ def test_tqdm_progress_bar_can_be_pickled():
...
+    total_val_batches = limit_batches * val_checks_per_epoch
+
+    assert pbar_callback.val_progress_bar.n == limit_batches
+    assert pbar_callback.val_progress_bar.total == limit_batches
+    assert pbar_callback.main_progress_bar.n == limit_batches + total_val_batches
+    assert pbar_callback.main_progress_bar.total == limit_batches + total_val_batches
+    assert pbar_callback.is_enabled
+
+
+@RunIf(min_gpus=2, standalone=True)
```
Comment (carmocca) [inline tests/callbacks/test_tqdm_progress_bar.py:594]: I know you didn't add it, but why is this a gpu standalone test?

Pbar doesn't have any distributed specific logic...

File: pytorch_lightning/callbacks/progress/tqdm_progress.py (hunk 4 above)
Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress/tqdm_progress.py:321]: ideally no.. since
```py
self.test_progress_bar.total == self.total_test_batches_current_dataloader
```
since I had to change it for the main progress bar, changed then here too.

But yes rich has some issues currently in some edge cases. Actually, I started with Rich, and ended up fixing tqdm.. there will be a follow-up for rich

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 5 above)
Comment (rohitgr7) [reply tests/callbacks/test_tqdm_progress_bar.py:350]: yes, so since `train batches = 1` and `val batches = 1` and refresh rate is 2, so ideally main progress bar should only update when total progress is 2 since it contains train+val.

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 6 above)
Comment (rohitgr7) [reply tests/callbacks/test_tqdm_progress_bar.py:353]: same here. `train batches = 5` and `val batches = 2` thus `total = 7` and refresh_rate=3, so for main progress bar updates should happen with multiples of 3 i.e 3, 6, 7(this is the end)

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 7 above)
Comment (rohitgr7) [reply tests/callbacks/test_tqdm_progress_bar.py:594]: will remove it.. might have been added by mistake.
but the interesting thing is that this test was failing with `standalone=True` but not normally. #12474

will investigate more on this.

Comment (awaelchli) [reply tests/callbacks/test_tqdm_progress_bar.py:594]: If we follow git blame back, it was added here #7863. I guess the decorator and global_zero check was added because the source issue was reported for a multi-gpu/node case.

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 5 above)
Comment (carmocca) [reply tests/callbacks/test_tqdm_progress_bar.py:350]: I disagree here. The refresh rate is meant to reduce step-level updates but I don't think it should apply to epoch-level updates.

With this change, metrics logged on_epoch do not appear in the training progress and I think that's counterintuitive for the user, even if it's a stricter view of the refresh rate concept.

File: tests/callbacks/test_tqdm_progress_bar.py (hunk 7 above)
Comment (carmocca) [reply tests/callbacks/test_tqdm_progress_bar.py:594]: (Can be done in a follow-up) I still suggest we remove it from this test. Standalone tests are the current bottleneck in CI

**Review by carmocca — APPROVED** [review]
LGTM

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -82,6 +82,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
+- Run main progress bar independent of val progress bar in `TQDMProgressBar` ([#12563](https://github.com/PyTorchLightning/pytorch-lightning/pull/12563))
```
Comment (awaelchli) [inline CHANGELOG.md:85]: this could be misinterpreted. the two bars are not independent. the main bar always contains the count of all validation steps. 

```suggestion
- Run main progress bar updates independent of val progress bar updates in `TQDMProgressBar` ([#12563](https://github.com/PyTorchLightning/pytorch-lightning/pull/12563))
```
